### ISSUE TITLE
Core signatures need improvement

### ISSUE THREAD

**Issue by dsyme** [fsprojects/fsharp-core-api-docs#8]
This is too complex, we can show this information in a simpler way:

 
![image](https://user-images.githubusercontent.com/7204669/87593896-0b246880-c6e4-11ea-8911-fe972268cc2d.png)

Possible improvements 

1. Show `List.allPairs` not `allPairs` because this is RequireQualifiedAccess module

2. Consider automatically showing example usage e.g. `List.allPairs list1 list2`

3. Show types of each parameter. e.g.

       List.allPairs list1 list2
           list1: 'T1 list
           list2: 'T2 list
           returns: ('T1 * 'T2) list

4. Add hover tips for each of `allPairs`, `list1`, `list2` showing full signature

5. Make examples, exceptions etc. are shown in the docs

**Comment by baronfel** [comment]
The signatures themselves are currently shown just by passing through the F#F signatures from `ApiDocMember.SignatureTooltip`, so either these tips would need to change (possible with formatting flags to control layout?) or we'd need to start constructing these signatures from member arguments and return types in a more manual manner.

**Comment by dsyme** [comment]
> The signatures themselves are currently shown just by passing through the F# signatures from ApiDocMember.SignatureTooltip, so either these tips would need to change (possible with formatting flags to control layout?) 

Yes.  There is a long journey to make this "right" and "simple" and "beautiful" which is why I expect so much churn here (among other things)

Latest FSharp.Formatting gives `ApiDocMember.FullUsageTooltip` and a few others so the default doc builder constructs this (I'm not saying this is right, but the elements are there).

There is also the actual tooltip formatting that flows out of the F# Compiler Service we could use, that gives colorization information as well.  At the moment we're only picking up the plaintext.

![image](https://user-images.githubusercontent.com/7204669/87881496-7d04f680-c9f1-11ea-9ddc-d2782df0fa5d.png)


> or we'd need to start constructing these signatures from member arguments and return types in a more manual manner.

Let's not do this because whatever we do will be useful for every F# project.

**Comment by Krzysztof-Cieslak** [comment]
As a note - Saturn documentation has changed the way we display signatures while ago by putting signature into tooltip as in old F#.Formatting template (example - https://saturnframework.org/reference/Saturn/saturn-pipelinehelpers.html) - it's still using older F#.Formatting so it's not as nice as thing shown by Don above but code is there

> we'd need to start constructing these signatures from member arguments and return types in a more manual manner.
> Let's not do this because whatever we do will be useful for every F# project.

I think Chet means constructing tooltips in a similar manner FSAC does (or even using same code, it should be fairly easily portable) - Ionide's tooltip and Info Panel is still way ahead of anything else in the ecosystem, both in how it looks and how much content it includes
![tooltip-Ionide](https://user-images.githubusercontent.com/5427083/87881857-63b57780-c9fc-11ea-949b-a5f7cecfec75.gif)

Repeating content is probably not necessary, but nicer formatting of signature, full name, etc can be useful.

**Comment by baronfel** [comment]
Yes, Krzystof did a better job of explaining what I meant

**Comment by dsyme** [comment]
This look great

One small concern is that by default these large displays this will require templates to have scrollbar support in the pages for these popups.  It might be good to make that optional

As an aside, do you have any code to crack the parameter info out of markdown comments?    (btw I'm wondering if we should just bite the bullert and put support for markdown comments into the F# compiler, or at least FCS)

**Comment by Krzysztof-Cieslak** [comment]
> btw I'm wondering if we should just bite the bullert and put support for markdown comments into the F# compiler, or at least FCS

Well, that's what I've tried to push forward with Ionide having comment to generate opinionated markdown commented while ago - we were forced to reverse that change as no other IDE supports displaying markdown comments nicely, and we now generate normal XML comments instead 🙄

**Comment by Krzysztof-Cieslak** [comment]
> Scrollbar 

I think it may not be actually a big issue if we don't put content (summary etc) of comment into the tooltip (what Ionide obviously does in the GIF above- but it's not necessary for a webpage, the content is there in next table cell) but just nicely formatted signature and some simple additional things (full name, declaring entity )

**Comment by dsyme** [comment]
> Well, that's what I've tried to push forward with Ionide having comment to generate opinionated markdown commented while ago - we were forced to reverse that change as no other IDE supports displaying markdown comments nicely, and we now generate normal XML comments instead 🙄

Worth discussing with @cartermp .  Hard to deviate from norms on something so core, but if the support was in the compiler to transform markdown --> XML  then it would be plausible.   The .XML files would still contain XML comments however

**Comment by cartermp** [comment]
Post-F# 5 this will be a very real consideration for F# tools given our eventual goal of centralizing things.

**Comment by dsyme** [comment]
@cartermp Do you think F# can afford to go it alone on a new interpretation of `///` as markdown?  There may be a compiler switch needed.... Also what if C# later decides to add the same feature?

**Comment by cartermp** [comment]
Considering that we already process XML docs in our own way i don't think C# is much of a consideration for the processing of markdown files. I would rather see this show up in tooling first and then have it works its way into a compiler layer over time as it is hardened though

**Comment by Krzysztof-Cieslak** [comment]
Just for sake of dicsussion what we proposed in Ionide was:
```
    /// **Description**
    ///
    /// **Parameters**
    ///   * `person` - parameter of type `Person`
    ///
    /// **Output Type**
    ///   * `string`
    ///
    /// **Exceptions**
    ///
```

I can imagine parsing sections like `Parameters` or `Exceptions` into appropriate XMLDoc tags.

**Comment by dsyme** [comment]
@Krzysztof-Cieslak What's the most promising way to get a hosted hierarchical menu like this on the client side, and what information do we need to emit from `ApiDocs.GenerateModel` and `ApiDocs.GenerateHtml` to get this?

![image](https://user-images.githubusercontent.com/7204669/87963696-08918c80-cab1-11ea-801d-191b6684fc66.png)

**Comment by dsyme** [comment]
> Considering that we already process XML docs in our own way i don't think C# is much of a consideration for the processing of markdown files. 

To be clear, the proposal is about how `///` comments are interpreted by F# compiler tooling.  That is

    /// some-special-markdown

(with a spec like this https://github.com/fsharp/fsharp-core-api-docs/issues/8#issuecomment-661111735) would be emitted by the F# compiler into an .XML file as XML, not markdown.  It would also activate correctly in parameter help and IDE tooling.

>  I would rather see this show up in tooling first and then have it works its way into a compiler layer over time as it is hardened though

I don't think this is the right way around (and it is the approach that's led to half-baked support for markdown comments in the tooling eco system so far.)

The root implementation would be in the F# compiler and 

1. convert the `/// <some-special-markdown>` and `/// <some-xml>` into the same internal structure
2. check that structure in both cases
3. emit it as XML  doc files
4. make the elements available in FCS, e.g. for FSHarp.Formatting and other documentation generation tools

Anything built above the F# compiler layer will not really work, though we could prototype some elements of it (basically, anything built on FSharp.Formatting already has some support half-baked for it).

**Comment by Krzysztof-Cieslak** [comment]
The problem is that if we will give people markdown... then well, they’ll use markdown. We can parse comment, split it into parts and put content into appropriate XMLDoc tags but the content inside those tags will still be markdown - it will include markdown way of bold/italics/underscore, markdown way of creating lists, markdown way of defining links etc. 
We need to make sure it’s displayed nicely in VS (which was a huge issue, when some project adopted what I proposed) - because markdown will work fine in Ionide and with F#.Formatting but nowhere else.

**Comment by dsyme** [comment]
Right.  Either VS understands those markdown comments or it doesn't . Without that we split the universe, including for all consumers of a component.

**Comment by cartermp** [comment]
> To be clear, the proposal is about how /// comments are interpreted by F# compiler tooling.

So am I. We process these files completely independently from how C# does, aside from the fact that we're literally emitting or reading XML. So when you ask this:

> Do you think F# can afford to go it alone on a new interpretation of /// as markdown? 

Yes, since we're already doing that with how we interpret `///` and XML docs.

> I don't think this is the right way around (and it is the approach that's led to half-baked support for markdown comments in the tooling eco system so far.)

Ehhh, we've seen enough examples of doing things in FSharp.Editor and moving them into FCS once they're stable before. It's actually been pretty successful: some core analysis undergoes churn for a few VS releases until it stabilizes, then moves into FCS so everyone gets something stable. The process of tagging markdown-embedded-in-XML so that a rendering system can do stuff with it should ultimately live in FCS but could undergo some churn, especially if we choose to do what C# does and associate a type in XML docs with a symbol the compiler can resolve. That would let us better distinguish different things in a rendering system, which we don't really do at all today.

To interpret markdown in XML docs is probably not too bad, but we'd need to create our own WPF views for VS tooltips similar to how we render links for navigation. Except it would be much more extensive than just rendering links.

Alternatively we do nothing until some unknown time in the future when VS uses LSP for everything, which should give us markdown rendering "for free".

**Comment by voronoipotato** [comment]
I mean the do nothing option just means having markdown syntax visible to the user, which is the present context of using the xml. I think in that way "doing nothing" is probably fine at least in the near term. Most developers are capable of reading and mentally interpreting markdown, and if they're not they will be after a few tries.

**Comment by baronfel** [comment]
@voronoipotato  It's a bit of a tooling problem, though, because tools use things like the `<param>` element to give parameter-specific tooltips when you're on that parameter's 'space' in a method or function call. That's the kind of 'free' integration that you'd miss if there wasn't a fully-fleshed out markdown integration, and it's precisely this terrible middle ground that led to Maxime rewriting FSAC's xml parser handler to do all the right things, leading the way for us to remove the markdown support entirely. I don't think Ionide is such a big player that we can just use force of install numbers to sway VS/VSForMac/Rider in this way.

**Comment by dsyme** [comment]
>  We process everything completely independently from how C# does, aside from the fact that we're literally emitting or reading XML. 

I don't think this is the case? Here's my understanding:

* The F# compiler emits XML files that C# IDE tooling and generic .NET documentation tools can accept, for example.  The FSharp.Core.xml is one example of this.  So when someone uses documentation comments in F# and delivers the .XML file in a nuget package they are seen by C#. XML docs are a .NET standard.  

* F# doesn't check the docs at compile-time like C# does, and also allows `///` comments without `<summary></summary`, but otherwise I believe our processing is the same in the sense that we generate the XML

* Within F# assemblies in the IDE is a different story.  The VS IDE tools do not correctly interpret `///` docs as proper XML docs for in-assembly references but make a half-assed attempt to convert the docs into usable quick info. They also allow `///` in many places where there is no corresponding entry in the XML file.
